crates/sparrow-journal/src/checkpoints.rs

```diff
@@ -0,0 +1,254 @@
+use futures::{StreamExt, TryStreamExt};
+use sparrow_batch::{Batch, RowTime};
+
+use crate::error::Error;
+
+/// Manages the checkpointed files.
+///
+/// We could use Parquet files for check-pointing but we instead use
+/// Arrow IPC files, since we believe they should be easier to read.
+///
+/// TODO: Support checkpoints on S3.
+#[derive(Debug)]
+pub(super) struct Checkpoints {
+    checkpoint_dir: PathBuf,
+    checkpoints: Vec<CheckpointFile>,
+    schema: SchemaRef,
+}
+
+#[derive(Debug)]
+struct CheckpointFile {
+    file_name: String,
+    min_present_time: RowTime,
+    max_present_time: RowTime,
+}
+
+/// Restore metadata of checkpoints by scanning entries in the directory.
+///
+/// TODO: In the future, we should store checkpoint metadata in the state DB
+/// and do this recovery from that.
+fn recover_checkpoints(
+    checkpoint_path: &std::path::Path,
+    schema: &SchemaRef,
+) -> error_stack::Result<Vec<CheckpointFile>, Error> {
+    let mut checkpoints = vec![];
+    if checkpoint_path.exists() {
+        error_stack::ensure!(
+            checkpoint_path.is_dir(),
+            Error::CheckpointDirNotDir(checkpoint_path.to_path_buf())
+        );
+    } else {
+        std::fs::create_dir_all(checkpoint_path)
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+    }
+
+    for file in std::fs::read_dir(checkpoint_path)
+        .into_report()
+        .change_context(Error::RecoveringCheckpoints)?
+    {
+        let file = file
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+        let file_type = file
+            .file_type()
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+        if !file_type.is_file() || file.path().extension() != Some("arrow".as_ref()) {
+            continue;
+        }
+
+        let reader = std::fs::File::open(file.path())
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+        let reader = arrow_ipc::reader::FileReader::try_new(reader, None)
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+        error_stack::ensure!(&reader.schema() == schema, Error::RecoveringCheckpoints);
+
+        let file_name = file
+            .file_name()
+            .into_string()
+            .map_err(|_| Error::RecoveringCheckpoints)?;
+
+        let min_present_time = reader
+            .custom_metadata()
+            .get("min_present_time")
+            .ok_or(Error::RecoveringCheckpoints)?;
+        let min_present_time =
+            RowTime::from_str(min_present_time).change_context(Error::RecoveringCheckpoints)?;
+
+        let max_present_time = reader
+            .custom_metadata()
+            .get("max_present_time")
+            .ok_or(Error::RecoveringCheckpoints)?;
+        let max_present_time =
+            RowTime::from_str(max_present_time).change_context(Error::RecoveringCheckpoints)?;
```

Comment:
Sheesh. Wonder if there would be a macro invocation or splitting this up into a helper or two to decrease the amount of `into_report().change_context(Error::RecoveringCheckpoint)`s